wemix/governance-contract/common/simulated-backend/contract.go

```diff
@@ -0,0 +1,263 @@
+	input := p.PackInput(t, method, args...)
+
+	msg := ethereum.CallMsg{From: common.Address{}, To: &p.Address, Data: input}
+	data, err := p.Client.Backend.CallContract(ctx, msg, nil)
+
+	require.NoErrorf(t, err, "call contract: %v", method)
+
+	output = reflect.New(outType).Interface()
+	if err = p.Abi.UnpackIntoInterface(&output, method, data); err != nil {
+		err = p.Abi.UnpackIntoInterface(output, method, data)
+	}
+	if err != nil && p.Logic != nil {
+		if err = p.Logic.Abi.UnpackIntoInterface(&output, method, data); err != nil {
+			err = p.Logic.Abi.UnpackIntoInterface(output, method, data)
+		}
+	}
+	assert.NoError(t, err, "unpack values:%v", method)
+
+	return
+}
+
+// ExecuteOk is that must be successful.
+func (p *Contract) ExecuteOk(t *testing.T, sender common.Address, method string, args ...interface{}) (receipt *types.Receipt) {
+	return p.ExecuteWithETHOk(t, sender, big.NewInt(0), method, args...)
+}
+
+// ExecuteFail is that must be failed.
+func (p *Contract) ExecuteFail(t *testing.T, sender common.Address, method string, args ...interface{}) (err error) {
+	return p.ExecuteWithETHFail(t, sender, big.NewInt(0), method, args...)
+}
+
+func (p *Contract) ExecuteWithETHOk(t *testing.T, sender common.Address, value *big.Int, method string, args ...interface{}) (receipt *types.Receipt) {
+	var err error
+	receipt, err = p.execute(t, sender, value, method, args...)
+	require.Truef(t, receipt.Status == types.ReceiptStatusSuccessful, "execute error(%v): contract:%v, method:%v, gas:%v\n%v", receipt.Status, p.Alias, method, receipt.GasUsed, err)
+	return
+}
+
+func (p *Contract) ExecuteWithETHFail(t *testing.T, sender common.Address, value *big.Int, method string, args ...interface{}) (err error) {
+	var receipt *types.Receipt
+	receipt, err = p.execute(t, sender, value, method, args...)
+	require.False(t, receipt.Status == types.ReceiptStatusSuccessful, fmt.Sprintf("execute not error(%v): contract:%v, method:%v, gas:%v", receipt.Status, p.Alias, method, receipt.GasUsed))
+	return
+}
+
+// ExecuteWithKlay includes value in the execution transaction.
+func (p *Contract) execute(t *testing.T, sender common.Address, value *big.Int, method string, args ...interface{}) (receipt *types.Receipt, err error) {
+	var tx *types.Transaction
+	if sender == (common.Address{}) {
+		sender = p.Client.Owner
+	}
+
+	tx, receipt = p.Client.SendTransaction(t, sender, p.Address, value, p.PackInput(t, method, args...))
+	if receipt.Status != types.ReceiptStatusSuccessful {
+		err = CallRevertReason(t, p.Client.Backend, sender, receipt.BlockNumber, tx)
+	} else {
+		if p.Logic != nil {
+			if e, err := p.Abi.EventByID(crypto.Keccak256Hash([]byte("Upgraded(address)"))); err == nil {
+			search:
+				for _, g := range receipt.Logs {
+					if g.Topics[0] == e.ID {
+						newLogic := p.Client.FindContractByAddress(common.BytesToAddress(g.Topics[1][:]))
+						require.NotNil(t, newLogic)
+						p.Logic = newLogic
+						break search
+					}
+				}
+			}
+		}
```

Comment:
check presence if an Upgraded event has been occurred.

```suggestion
			if p.Logic != nil {
		topic := crypto.Keccak256Hash([]byte("Upgraded(address)"))
		if types.BloomLookup(receipt.Bloom, topic) {
			if e, err := p.Abi.EventByID(topic); err == nil {
				for _, g := range receipt.Logs {
					if g.Topics[0] == e.ID {
						newLogic := p.Client.FindContractByAddress(common.BytesToAddress(g.Topics[1][:]))
						require.NotNil(t, newLogic)
						p.Logic = newLogic
						break
					}
				}
			}
		}
	}
```